File: 프로그래머스/level1/신고 결과 받기.cpp
//프로그래머스 신고 결과 받기
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <sstream>

using namespace std;

vector<int> solution(vector<string> id_list, vector<string> report, int k) {
    vector<int> answer;
    unordered_map<string, unordered_map<string, int>> um;
    unordered_map<string, int> report_id;

    for (int i = 0; i < report.size(); i++) {
        stringstream ss(report[i]);
        string s1, s2;
        ss >> s1 >> s2;

        um[s1][s2]++;
        if (um[s1][s2] == 1) {
            report_id[s2]++; //한 유저가 처음 신고한 경우
        }
    }

    for (int i = 0; i < id_list.size(); i++) {
        int num = 0;
        for (auto m : um[id_list[i]]) {
            if (report_id[m.first] >= k) {
                num++;
            }
        }
        answer.push_back(num);
    }

    return answer;
}

int main() {
    vector<string> id_list = { "con", "ryan" };
    vector<string> report = { "ryan con", "ryan con", "ryan con", "ryan con" };
    int k = 3;

    solution(id_list, report, k);

}

File: 프로그래머스(JAVA)/level1/신고결과받기.java
import java.util.*;

public class 신고결과받기 {
    static class Solution {
        public static int[] solution(String[] id_list, String[] report, int k) {
            int[] answer = new int[id_list.length];
            HashMap<String, List<String>> hashmap = new HashMap<>();
            HashMap<String, Integer> report_id = new HashMap<>();
            for(int i=0; i< id_list.length; i++){
                List<String> l = new LinkedList<>();
                hashmap.put(id_list[i], l);
            }
            for(int i=0; i< report.length; i++) {
                String[] s = report[i].split(" ");
                if(hashmap.get(s[0]).contains(s[1]) == false){
                    hashmap.get(s[0]).add(s[1]);
                    if(report_id.containsKey(s[1]) == false){
                        report_id.put(s[1], 1);
                    }else {
                        report_id.put(s[1], report_id.get(s[1]) + 1);
                    }
                }

            }

            for(int i=0; i< id_list.length; i++){
                Integer num = 0;
                for(String s : hashmap.get(id_list[i])){
                    if(report_id.get(s) >= k){
                        num++;
                    }
                }
                answer[i] = num;
            }

            return answer;
        }
        public static void main(String[] args){
            String[] id_list = {"muzi", "frodo", "apeach", "neo"};
            String[] report = {"muzi frodo","apeach frodo","frodo neo","muzi neo","apeach muzi"};
            int k = 2;
            solution(id_list, report,k);
        }
    }

}
